File: src/Service/Service.java
package Service;

import Models.*;
import Repos.*;

import javax.print.CancelablePrintJob;
import java.sql.Statement;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.List;

public class Service {
    private Connection connection;
    private final Repository<CellarLocation> cellarLocationRepo;
    private final Repository<Inventory> inventoryRepo;
    private final Repository<Region> regionrepo;
    private final Repository<Supplier> supplierRepo;
    private final Repository<Transactions> transactionsRepo;
    private final Repository<Users> userrepo;
    private final Repository<Wines> winerepo;
    private final Repository<WineType> wineTypeRepo;

    public Service(String dbUrl) throws Exception {
        this.connection= DriverManager.getConnection(dbUrl);
        this.cellarLocationRepo = new DBCellarLocationRepo(dbUrl);
        this.inventoryRepo = new DBInventoryRepo(dbUrl);
        this.regionrepo = new DBRegionRepo(dbUrl);
        this.supplierRepo = new DBSupplierRepo(dbUrl);
        this.transactionsRepo = new DBTransactionsRepo(dbUrl);
        this.userrepo = new DBUsersRepo(dbUrl);
        this.winerepo = new DBWinesRepo(dbUrl);
        this.wineTypeRepo = new DBWineTypeRepo(dbUrl);
    }

    public Connection getConnection(String dbUrl) throws SQLException {
        return DriverManager.getConnection(dbUrl);
    }


    public void createWine(Wines wine){
        winerepo.create(wine);
    }

    public Wines getWine(int id){
        return winerepo.get(id);
    }

        public void updateWine(int wine_id, String newName, int newTypeID, int newRegionID,int  newVintageYear, int alcoholPercentage, int supplierID) {
            Wines wine = winerepo.get(wine_id);
            if (wine == null) {
                System.out.println("Wine with ID " + wine_id + " not found.");
                return;
            }
            wine.setName(newName);
            wine.setType_id(newTypeID);
            wine.setRegion_id(newRegionID);
            wine.setVintage_year(newVintageYear);
            wine.setAlcohol_percentage(alcoholPercentage);
            wine.setSupplier_id(supplierID);
            winerepo.update(wine);
            System.out.println("Wine updated successfully.");
        }

    public void deleteWine(int id){
        winerepo.delete(id);
    }

    public List<Wines> getAllWines(){
        return winerepo.getAll();
    }

    public void createWineType(WineType wineType){
        wineTypeRepo.create(wineType);
    }

    public WineType getWineType(int id){
        return wineTypeRepo.get(id);
    }

    public void updateWineType(int winetype_id, String newName) {
        WineType wineType = wineTypeRepo.get(winetype_id);
        if (wineType == null) {
            System.out.println("WineType with ID " + winetype_id + " not found.");
            return;
        }
        wineType.setName(newName);
        wineTypeRepo.update(wineType);
        System.out.println("WineType updated successfully.");
        
    }

    public void deleteWineType(int id){
        wineTypeRepo.delete(id);
    }

    public List<WineType> getAllWineTypes(){
        return wineTypeRepo.getAll();
    }

    public void createRegion(Region region){
        regionrepo.create(region);
    }

    public Region getRegion(int id){
        return regionrepo.get(id);
    }

    public void updateRegion(int region_id, String newRegionName, String newRegionCountry) {
        Region region = regionrepo.get(region_id);
        if (region == null) {
            System.out.println("Region with ID " + region_id + " not found.");
            return;
        }
        region.setRegion_name(newRegionName);
        region.setCountry(newRegionCountry);
        regionrepo.update(region);
        System.out.println("Region updated successfully.");

    }

    public void deleteRegion(int id){
        regionrepo.delete(id);
    }

    public List<Region> getAllRegions(){
        return regionrepo.getAll();
    }

    public void createSupplier(Supplier supplier){
        supplierRepo.create(supplier);
    }

    public Supplier getSupplier(int id){
        return supplierRepo.get(id);
    }

    public void updateSupplier(int supplier_id, String newName,String newEmail, String newPhone,String newAddress) {
        Supplier supplier = supplierRepo.get(supplier_id);
        if (supplier == null) {
            System.out.println("Supplier with ID " + supplier_id + " not found.");
            return;
        }
        supplier.setSupplier_name(newName);
        supplier.setAddress(newAddress);
        supplier.setPhone(newPhone);
        supplier.setContact_email(newEmail);
        supplierRepo.update(supplier);
        System.out.println("Supplier updated successfully.");

    }

    public void deleteSupplier(int id){
        supplierRepo.delete(id);
    }

    public List<Supplier> getAllSuppliers(){
        return supplierRepo.getAll();
    }

    public void createCellarLocation(CellarLocation cellarLocation){
        cellarLocationRepo.create(cellarLocation);
    }

    public CellarLocation getCellarLocation(int id){
        return cellarLocationRepo.get(id);
    }

    public void updateCellarLocation(int location_id, String newSection, int newRackNumber, int newBottlePosition) {
        CellarLocation cellarLocation = cellarLocationRepo.get(location_id);
        if (cellarLocation == null) {
            System.out.println("CellarLocation with ID " + location_id + " not found.");
            return;
        }
        cellarLocation.setSection(newSection);
        cellarLocation.setRack_number(newRackNumber);
        cellarLocation.setBottle_position(newBottlePosition);
        cellarLocationRepo.update(cellarLocation);
        System.out.println("CellarLocation updated successfully.");

    }

    public void deleteCellarLocation(int id){
        cellarLocationRepo.delete(id);
    }

    public List<CellarLocation> getAllCellarLocations(){
        return cellarLocationRepo.getAll();
    }

    public void createInventory(Inventory inventory){
        inventoryRepo.create(inventory);
    }

    public Inventory getInventory(int id){
        return inventoryRepo.get(id);
    }

    public void updateInventory(int inventory_id, int newWine_id, int newLocation_id, int newQuantity, int newBottle_size_ml) {
        Inventory inventory = inventoryRepo.get(inventory_id);
        if (inventory == null) {
            System.out.println("Inventory with ID " + inventory_id + " not found.");
            return;
        }
        inventory.setWine_id(newWine_id);
        inventory.setLocation_id(newLocation_id);
        inventory.setQuantity(newQuantity);
        inventory.setBottle_size_ml(newBottle_size_ml);
        inventoryRepo.update(inventory);
        System.out.println("Inventory updated successfully.");
    }

    public void deleteInventory(int id){
        inventoryRepo.delete(id);
    }

    public List<Inventory> getAllInventories(){
        return inventoryRepo.getAll();
    }

    public void createTransaction(Transactions transaction){
        transactionsRepo.create(transaction);
    }

    public Transactions getTransaction(int id){
        return transactionsRepo.get(id);
    }

    public void updateTransaction(int transaction_id, int newWine_id, String newTransactionType, String newDate, int newQuantity, int newPricePerBottle, int newUserID) {
        Transactions transaction = transactionsRepo.get(transaction_id);
        if (transaction == null) {
            System.out.println("Transaction with ID " + transaction_id + " not found.");
            return;
        }
        transaction.setWine_id(newWine_id);
        transaction.setTransaction_type(newTransactionType);
        transaction.setDate(newDate);
        transaction.setQuantity(newQuantity);
        transaction.setPrice_per_bottle(newPricePerBottle);
        transaction.setUser_id(newUserID);
        transactionsRepo.update(transaction);
        System.out.println("Transaction updated successfully.");

    }

    public void deleteTransaction(int id){
        transactionsRepo.delete(id);
    }

    public List<Transactions> getAllTransactions(){
        return transactionsRepo.getAll();
    }

    public void createUser(Users user){
        userrepo.create(user);
    }

    public Users getUser(int id){
        return userrepo.get(id);
    }

    public void updateUser(int user_id, String newName, String newPasswordHash, String newRole) {
        Users user = userrepo.get(user_id);
        if (user == null) {
            System.out.println("User with ID " + user_id + " not found.");
            return;
        }
        user.setUsername(newName);
        user.setPassword_hash(newPasswordHash);
        user.setRole(newRole);
        userrepo.update(user);
        System.out.println("User updated successfully.");

    }

    public void deleteUser(int id){
        userrepo.delete(id);
    }

    public List<Users> getAllUsers(){
        return userrepo.getAll();
    }

    public void executeStatement(String sql) {
        try {
            Statement stmt = connection.createStatement();
            stmt.execute(sql);
            System.out.println("Executed: " + sql);
            stmt.close();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

}

File: Python code/Service.py
from Entities.CellarLocation import CellarLocation
from Entities.Inventory import Inventory
from Entities.Region import Region
from Entities.Supplier import Supplier
from Entities.Transaction import Transaction
from Entities.Users import Users
from Entities.Wines import Wines
from Entities.Wine_Type import Wine_Type
from Repos.CellarLocationRepo import CellarLocationRepo
from Repos.InventoryRepo import InventoryRepo
from Repos.RegionRepo import RegionRepo
from Repos.SupplierRepo import SupplierRepo
from Repos.TransactionRepo import TransactionRepo
from Repos.UserRepo import UserRepo
from Repos.WinesRepo import WinesRepo
from Repos.Wine_TypeRepo import Wine_TypeRepo

class Service:
    def __init__(self, ssms_conn, pg_conn):
        self.cellar_location_repo = CellarLocationRepo(ssms_conn, pg_conn)
        self.inventory_repo = InventoryRepo(ssms_conn, pg_conn)
        self.region_repo = RegionRepo(ssms_conn, pg_conn)
        self.supplier_repo = SupplierRepo(ssms_conn, pg_conn)
        self.transaction_repo = TransactionRepo(ssms_conn, pg_conn)
        self.user_repo = UserRepo(ssms_conn, pg_conn)
        self.wine_repo = WinesRepo(ssms_conn, pg_conn)
        self.wine_type_repo = Wine_TypeRepo(ssms_conn, pg_conn)

    def create_cellar_location(self, location_id, section, rack_number, bottle_position):
        cellar_location = CellarLocation(location_id,section,rack_number,bottle_position)
        self.cellar_location_repo.create_cellarlocation(cellar_location)

    def read_cellar_locations(self):
        self.cellar_location_repo.read_cellarlocations()

    def update_cellar_location(self, location_id, section, rack_number, bottle_position):
        cellar_location = CellarLocation(location_id,section,rack_number,bottle_position)
        self.cellar_location_repo.update_cellarlocation(cellar_location)

    def delete_cellar_location(self, location_id):
        self.cellar_location_repo.delete_cellarlocation(location_id)

    def get_cellar_location(self, location_id):
        self.cellar_location_repo.get_cellarlocation(location_id)

    def create_inventory(self, inventory_id, wine_id, location_id, quantity, bottle_size_ml):
        inventory = Inventory(inventory_id, wine_id, location_id, quantity, bottle_size_ml)
        self.inventory_repo.create_inventory(inventory)

    def read_inventory(self):
        self.inventory_repo.read_inventory()

    def update_inventory(self, inventory_id, wine_id, location_id, quantity, bottle_size_ml):
        inventory = Inventory(inventory_id, wine_id, location_id, quantity, bottle_size_ml)
        self.inventory_repo.update_inventory(inventory)

    def delete_inventory(self, inventory_id):
        self.inventory_repo.delete_inventory(inventory_id)

    def get_inventory(self, inventory_id):
        self.inventory_repo.get_inventory(inventory_id)

    def create_region(self, region_id, region_name,country):
        region=Region(region_id, region_name,country)
        self.region_repo.create_region(region)

    def read_region(self):
        self.region_repo.read_region()

    def update_region(self, region_id, region_name,country):
        region=Region(region_id, region_name,country)
        self.region_repo.update_region(region)

    def delete_region(self, region_id):
        self.region_repo.delete_region(region_id)

    def get_region(self, region_id):
        self.region_repo.get_region(region_id)

    def create_supplier(self, supplier_id, supplier_name, supplier_email, supplier_phone, supplier_address):
        supplier = Supplier(supplier_id, supplier_name, supplier_email, supplier_phone, supplier_address)
        self.supplier_repo.create_supplier(supplier)

    def read_supplier(self):
        self.supplier_repo.read_supplier()

    def update_supplier(self, supplier_id, supplier_name, supplier_email, supplier_phone, supplier_address):
        supplier = Supplier(supplier_id, supplier_name, supplier_email, supplier_phone, supplier_address)
        self.supplier_repo.update_supplier(supplier)

    def delete_supplier(self, supplier_id):
        self.supplier_repo.delete_supplier(supplier_id)

    def get_supplier(self, supplier_id):
        self.supplier_repo.get_supplier(supplier_id)

    def create_transaction(self, transaction_id, transaction_wine_id, transaction_type,transaction_date, transaction_quantity, price_per_bottle, user_id):
        transaction = Transaction(transaction_id, transaction_wine_id, transaction_type, transaction_date, transaction_quantity, price_per_bottle, user_id)
        self.transaction_repo.create_transaction(transaction)

    def read_transaction(self):
        self.transaction_repo.read_transactions()

    def update_transaction(self, transaction_id, transaction_wine_id, transaction_type,transaction_date, transaction_quantity, price_per_bottle, user_id):
        transaction = Transaction(transaction_id, transaction_wine_id, transaction_type, transaction_date, transaction_quantity, price_per_bottle, user_id)
        self.transaction_repo.update_transaction(transaction)

    def delete_transaction(self, transaction_id):
        self.transaction_repo.delete_transaction(transaction_id)

    def get_transaction(self, transaction_id):
        self.transaction_repo.get_transaction(transaction_id)

    def create_user(self, user_id, user_name, user_hash, user_role):
        user = Users(user_id, user_name, user_hash, user_role)
        self.user_repo.create_user(user)

    def read_user(self):
        self.user_repo.read_users()

    def update_user(self, user_id, user_name, user_hash, user_role):
        user = Users(user_id, user_name, user_hash, user_role)
        self.user_repo.update_user(user)

    def delete_user(self, user_id):
        self.user_repo.delete_user(user_id)

    def get_user(self, user_id):
        self.user_repo.get_user(user_id)

    def create_wine(self, wine_id, wine_name, wine_type_id, wine_region_id, vintage_year, alcohol_percentage, supplier_id):
        wine = Wines(wine_id, wine_name, wine_type_id, wine_region_id, vintage_year, alcohol_percentage, supplier_id)
        self.wine_repo.create_wine(wine)

    def read_wine(self):
        self.wine_repo.read_wines()

    def update_wine(self, wine_id, wine_name, wine_type_id, wine_region_id, vintage_year, alcohol_percentage, supplier_id):
        wine = Wines(wine_id, wine_name, wine_type_id, wine_region_id, vintage_year, alcohol_percentage, supplier_id)
        self.wine_repo.update_wine(wine)

    def delete_wine(self, wine_id):
        self.wine_repo.delete_wine(wine_id)

    def get_wine(self, wine_id):
        self.wine_repo.get_wine(wine_id)

    def create_wine_type(self, type_id, type_name):
        wine_type = Wine_Type(type_id, type_name)
        self.wine_type_repo.create_wine_type(wine_type)

    def read_wine_type(self):
        self.wine_type_repo.read_wine_types()

    def update_wine_type(self, type_id, type_name):
        wine_type = Wine_Type(type_id, type_name)
        self.wine_type_repo.update_wine_type(wine_type)

    def delete_wine_type(self, type_id):
        self.wine_type_repo.delete_wine_type(type_id)

    def get_wine_type(self, type_id):
        self.wine_type_repo.get_wine_type(type_id)








